Nodejs/Product/Nodejs/Intellisense/ExpansionClient.cs

```diff
@@ -0,0 +1,332 @@
+                && declarations != null) {
+                foreach (IXMLDOMNode declType in declarations.childNodes) {
+                    var id = declType.selectSingleNode("./CodeSnippet:ID");
+                    if (id != null) {
+                        declList.Add(id.text);
+                    }
+                }
+            }
+
+            List<string> importList = new List<string>();
+            if (ErrorHandler.Succeeded(hr = _session.GetSnippetNode("CodeSnippet:Imports", out imports))
+                && imports != null) {
+                foreach (IXMLDOMNode import in imports.childNodes) {
+                    var id = import.selectSingleNode("./CodeSnippet:Namespace");
+                    if (id != null) {
+                        importList.Add(id.text);
+                    }
+                }
+            }
+
+            bool surroundsWith = false, surroundsWithStatement = false;
+            foreach (MSXML.IXMLDOMNode snippetType in snippetTypes.childNodes) {
+                if (snippetType.nodeName == "SnippetType") {
+                    if (snippetType.text == SurroundsWith) {
+                        surroundsWith = true;
+                    } else if (snippetType.text == SurroundsWithStatement) {
+                        surroundsWithStatement = true;
+                    }
+                }
+            }
+            // get the indentation of where we're inserting the code...
+            string baseIndentation = GetBaseIndentation(ts);
+
+            TextSpan? endSpan = null;
+            using (var edit = _textView.TextBuffer.CreateEdit()) {
+                if (surroundsWith || surroundsWithStatement) {
+                    // this is super annoyning...  Most languages can do a surround with and $selected$ can be
+                    // an empty string and everything's the same.  But in Python we can't just have something like
+                    // "while True: " without a pass statement.  So if we start off with an empty selection we
+                    // need to insert a pass statement.  This is the purpose of the "SurroundsWithStatement"
+                    // snippet type.
+                    //
+                    // But, to make things even more complicated, we don't have a good indication of what's the 
+                    // template text vs. what's the selected text.  We do have access to the original template,
+                    // but all of the values have been replaced with their default values when we get called
+                    // here.  So we need to go back and re-apply the template, except for the $selected$ part.
+                    //
+                    // Also, the text has \n, but the inserted text has been replaced with the appropriate newline
+                    // character for the buffer.
+                    var templateText = codeNode.text.Replace("\n", _textView.Options.GetNewLineCharacter());

```

Comment:
The code snippets that you added are all CRLF (```\r\n```). They would need to be ```\n``` in order for this replacement to make sense.

Additionally, use ```VsExtensions.GetNewLineText``` instead of ```GetNewLineCharacter()```, which works around some annoying VS issues...
https://github.com/Microsoft/nodejstools/blob/master/Common/Product/SharedProject/VsExtensions.cs#L162